### PR TITLE
Add DINOv2 with registers

### LINKED ISSUE #27379: dinov2 with REGISTERS
### Model description

Dear huggingface team,

The fair team published an improved version of dinov2 [<VISION TRANSFORMERS NEED REGISTERS>](https://arxiv.org/abs/2309.16588). The models and checkpoints are available in the [dinov2 website](https://github.com/facebookresearch/dinov2#pretrained-backbones-via-pytorch-hub), but not in hugging face. 

Could you add this new model? I really appreciate your work.

Best Wishes,

Zongze

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

[<VISION TRANSFORMERS NEED REGISTERS>](https://arxiv.org/abs/2309.16588)
[dinov2 reg checkpoint](https://github.com/facebookresearch/dinov2#pretrained-backbones-via-pytorch-hub)

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds DINOv2 with registers, this time using the new modular tool.

Fixes #27379

To do:

- [x] make @bernardzach a co-author

### CODE REVIEW

File: src/transformers/models/dinov2_with_registers/__init__.py
Comment (ArthurZucker) [inline src/transformers/models/dinov2_with_registers/__init__.py:1]: this is the old __init__ !

**Review by ArthurZucker — APPROVED** [review]
Nice! The drop path update might be a bit breaking, you can import the function in modular IMO

File: src/transformers/models/dinov2/modeling_dinov2.py [hunk 1]
```diff
@@ -347,37 +347,33 @@ def forward(self, hidden_state: torch.Tensor) -> torch.Tensor:
...
-
-# Copied from transformers.models.beit.modeling_beit.BeitDropPath
 class Dinov2DropPath(nn.Module):
     """Drop paths (Stochastic Depth) per sample (when applied in main path of residual blocks)."""
 
     def __init__(self, drop_prob: Optional[float] = None) -> None:
         super().__init__()
         self.drop_prob = drop_prob
 
+    def drop_path(input: torch.Tensor, drop_prob: float = 0.0, training: bool = False) -> torch.Tensor:
```
Comment (qubvel) [inline src/transformers/models/dinov2/modeling_dinov2.py:357]:
```suggestion
    def drop_path(self, input: torch.Tensor, drop_prob: float = 0.0, training: bool = False) -> torch.Tensor:
```

**Comment by NielsRogge** [comment]
Ok cool looks like we can merge :)

All model checkpoints are on the hub, would need to be transferred to the facebook org: https://huggingface.co/models?other=dinov2_with_registers

**Review by ArthurZucker — COMMENTED** [review]
Yep feel free to merge in the meant time 🤗 thanks for updating !

**Comment by xenova** [comment]
@NielsRogge Would it be possible to adapt the `interpolate_pos_encoding` function to use the same approach as Dinov2 (but just with options for antialiasing):

https://github.com/huggingface/transformers/blob/6e0515e99c39444caae39472ee1b2fd76ece32f1/src/transformers/models/dinov2/modeling_dinov2.py#L78-L114

This current version has issues with tracing of the graph (see https://github.com/huggingface/transformers/pull/33226).

File: src/transformers/models/dinov2_with_registers/configuration_dinov2_with_registers.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    >>> configuration = Dinov2WithRegistersConfig()
+
+    >>> # Initializing a model (with random weights) from the base style configuration
+    >>> model = Dinov2WithRegistersModel(configuration)
+
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+
+    model_type = "dinov2-with-registers-base"
```
Comment (xenova) [inline src/transformers/models/dinov2_with_registers/configuration_dinov2_with_registers.py:108]: There is a mismatch between the models on the hub and the model_type specified here. See https://huggingface.co/facebook/dinov2-with-registers-small/blob/main/config.json#L18, it is `dinov2_with_registers`.

This causes an issue when loading the models and re-saving them (e.g., for finetuning or conversions)
```

Comment (NielsRogge) [reply src/transformers/models/dinov2_with_registers/configuration_dinov2_with_registers.py:108]: cc @ydshieh weird this wasn't caught by the tests, is this something we can add a test for?

**Pull request merged**